### PR TITLE
Refactored token handler (#8)

### PR DESCRIPTION
**Pull request by PKorovinPromo-Z**
* Api project added to solution

* Added ServiceCollectionExtensions

* Refactored TokenAuthHeaderHandler

* Fixed variable name

### CODE REVIEW

File: HappyTravel.FastpayhotelsConnector.Common/Constants.cs [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
+{   
     public const string HeaderGrandtype = "grant_type";
     public const string HeaderClientId = "client_id";
     public const string HeaderClientSecret = "client_secret";
     public const string HeaderVersion = "version";
     public const string HeaderUser = "username";
     public const string HeaderPassword = "password";
     public const string ApiVersion = "1";
+    public const string FastpayhotelsTokenClient = "FastpayhotelsTokenClient";
+    public static readonly TimeSpan AuthTokenLifeTime = TimeSpan.FromDays(400); // The access token has a validity of 400 days
```
Comment (suhan3z) [inline HappyTravel.FastpayhotelsConnector.Common/Constants.cs:13]: It's better to have a measurement scale in the name. AuthTokenLifeTimeDays for example

File: HappyTravel.FastpayhotelsConnector.Common/Infrastructure/Extensions/ServiceCollectionExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+using Microsoft.Extensions.DependencyInjection;
+using Polly;
+using Polly.Extensions.Http;
+using System.Net;
+
+namespace HappyTravel.FastpayhotelsConnector.Common.Infrastructure.Extensions
+{
+    public static class ServiceCollectionExtensions
+    {
+        public static IServiceCollection AddTokenAuthHeaderService(this IServiceCollection services, string tokenEndPoint)
```
Comment (suhan3z) [inline HappyTravel.FastpayhotelsConnector.Common/Infrastructure/Extensions/ServiceCollectionExtensions.cs:11]: You still challenging me by lack of comments. but i strong ;)

File: HappyTravel.FastpayhotelsConnector.Common/Infrastructure/Extensions/ServiceCollectionExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+                .AddPolicyHandler(GetDefaultRetryPolicy());
+
+            return services.AddTransient<TokenAuthHeaderClient>()
+                .AddTransient<TokenAuthHeaderService>()
+                .AddTransient<TokenAuthHeaderStorage>();
+
+
+            IAsyncPolicy<HttpResponseMessage> GetDefaultRetryPolicy()
+            {
+                var jitter = new Random();
```
Comment (suhan3z) [inline HappyTravel.FastpayhotelsConnector.Common/Infrastructure/Extensions/ServiceCollectionExtensions.cs:27]: What is jitter?

File: HappyTravel.FastpayhotelsConnector.Common/Infrastructure/Extensions/ServiceCollectionExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+{
+    public static class ServiceCollectionExtensions
+    {
+        public static IServiceCollection AddTokenAuthHeaderService(this IServiceCollection services, string tokenEndPoint)
+        {             
+            services.AddHttpClient(Constants.FastpayhotelsTokenClient, client =>
+                {
+                    client.BaseAddress = new Uri(tokenEndPoint);
+                })
+                .SetHandlerLifetime(TimeSpan.FromMinutes(5))
```
Comment (suhan3z) [inline HappyTravel.FastpayhotelsConnector.Common/Infrastructure/Extensions/ServiceCollectionExtensions.cs:17]: to const

File: HappyTravel.FastpayhotelsConnector.Common/Infrastructure/Extensions/ServiceCollectionExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+
+            IAsyncPolicy<HttpResponseMessage> GetDefaultRetryPolicy()
+            {
+                var jitter = new Random();
+
+                return HttpPolicyExtensions
+                    .HandleTransientHttpError()
+                    .OrResult(msg => msg.StatusCode == HttpStatusCode.ServiceUnavailable)
+                    .WaitAndRetryAsync(3, attempt
```
Comment (suhan3z) [inline HappyTravel.FastpayhotelsConnector.Common/Infrastructure/Extensions/ServiceCollectionExtensions.cs:32]: Too much magic numbers

File: HappyTravel.FastpayhotelsConnector.Common/Infrastructure/Extensions/ServiceCollectionExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+                        => TimeSpan.FromSeconds(Math.Pow(1.5, attempt)) + TimeSpan.FromMilliseconds(jitter.Next(0, 100)));
```
Comment (suhan3z) [inline HappyTravel.FastpayhotelsConnector.Common/Infrastructure/Extensions/ServiceCollectionExtensions.cs:33]: here too

File: HappyTravel.FastpayhotelsConnector.Common/Infrastructure/Options/ApiConnection.cs [hunk 7]
```diff
@@ -0,0 +1,12 @@
+﻿namespace HappyTravel.FastpayhotelsConnector.Common.Infrastructure.Options;
+
+public class ApiConnection
+{
+    public string AvailabilityEndPoint { get; set; }
```
Comment (suhan3z) [inline HappyTravel.FastpayhotelsConnector.Common/Infrastructure/Options/ApiConnection.cs:5]: Comments

File: HappyTravel.FastpayhotelsConnector.Common/Models/TokenErrorMessage.cs [hunk 8]
```diff
@@ -0,0 +1,12 @@
+﻿using System.Text.Json.Serialization;
+
+namespace HappyTravel.FastpayhotelsConnector.Common.Models;
+
+public class TokenErrorMessage
+{
+    [JsonPropertyName("error")]
+    public string Error { get; set; }
```
Comment (suhan3z) [inline HappyTravel.FastpayhotelsConnector.Common/Models/TokenErrorMessage.cs:8]: Comments

File: HappyTravel.FastpayhotelsConnector.Common/Service/TokenAuthHeader/TokenAuthHeaderClient.cs [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+public class TokenAuthHeaderClient
+{
+    public TokenAuthHeaderClient(IHttpClientFactory httpClientFactory, IOptions<ApiConnection> apiConnection)
+    {
+        _httpClientFactory = httpClientFactory;
+        _apiConnection = apiConnection.Value;
+    }
+
+
+    public async Task<Result<string>> GetAgencyRolToken(CancellationToken cancellationToken)
```
Comment (suhan3z) [inline HappyTravel.FastpayhotelsConnector.Common/Service/TokenAuthHeader/TokenAuthHeaderClient.cs:18]: Comments

File: HappyTravel.FastpayhotelsConnector.Common/Service/TokenAuthHeader/TokenAuthHeaderClient.cs [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+    {
+        const string endPoint = "/security/token";
+
+        var content = new FormUrlEncodedContent(new List<KeyValuePair<string, string>>()
```
Comment (suhan3z) [inline HappyTravel.FastpayhotelsConnector.Common/Service/TokenAuthHeader/TokenAuthHeaderClient.cs:22]: Maybe extract to service method?

File: HappyTravel.FastpayhotelsConnector.Common/Service/TokenAuthHeader/TokenAuthHeaderService.cs [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+{
+    public TokenAuthHeaderService(TokenAuthHeaderClient tokenAuthHeaderClient,
+        TokenAuthHeaderStorage tokenAuthHeaderStorage)
+    {
+        _tokenAuthHeaderClient = tokenAuthHeaderClient;
+        _tokenAuthHeaderStorage = tokenAuthHeaderStorage;
+    }
+
+
+    public async Task<Result<string>> GetOrSetToken(CancellationToken cancellationToken)
```
Comment (suhan3z) [inline HappyTravel.FastpayhotelsConnector.Common/Service/TokenAuthHeader/TokenAuthHeaderService.cs:15]: Very confusing method name and logic

File: HappyTravel.FastpayhotelsConnector.Common/Service/TokenAuthHeader/TokenAuthHeaderStorage.cs [hunk 12]
```diff
@@ -0,0 +1,35 @@
+﻿using FloxDc.CacheFlow;
+using FloxDc.CacheFlow.Extensions;
+
+namespace HappyTravel.FastpayhotelsConnector.Common.Service.TokenAuthHeader;
+
+public class TokenAuthHeaderStorage
```
Comment (suhan3z) [inline HappyTravel.FastpayhotelsConnector.Common/Service/TokenAuthHeader/TokenAuthHeaderStorage.cs:6]: Comments

**Review by suhan3z — CHANGES_REQUESTED** [review]
According to comments

File: HappyTravel.FastpayhotelsConnector.Common/Infrastructure/Extensions/ServiceCollectionExtensions.cs (hunk 3 above)
Comment (PKorovinPromo-Z) [reply HappyTravel.FastpayhotelsConnector.Common/Infrastructure/Extensions/ServiceCollectionExtensions.cs:27]: Taken from edo: https://github.com/happy-travel/edo/blob/469ce388215e0ddcc4cd07e23338d9911071b8ec/Api/Infrastructure/ServiceCollectionExtensions.cs#L824

**Review by suhan3z — APPROVED** [review]

**Pull request merged**